Start documenting Multisig SDKs (#616)
rsksmart/devportal: rif/multisig/sdk/transactions.md

layout: rsk
title: RIF Multisig SDK - transactions
---

**Required packages**:
- [safe-transactions-sdk](https://github.com/rsksmart/safe-transactions-sdk)
- [safe-core-sdk](https://github.com/gnosis/safe-core-sdk)

## Raw Transactions

Raw transactions can be created by using directly the [safe-core-sdk](https://github.com/gnosis/safe-core-sdk) package or the [safe-transactions-sdk](https://github.com/rsksmart/).

```
const tx = await safe.createTransaction({
      to,
      data,
      value
    })
```

### safe-transactions-sdk

The `RawTransactionBuilder` provides the possibility to create raw transactions.

Initialize the `RawTransactionBuilder`.

```
const rawTransactionBuilder = new RawTransactionBuilder(safe)
```

Create the transaction.

```
const tx = await rawTransactionBuilder.rawTransaction(to, value, data)
```

> ## ERC20 Transactions (including RIF Token)

Reviewer: Consider moving `(including RIF Token)` to the text of this section, instead of being in the section heading.

Reviewer 2: > LGTM
> 
>     * I have left some proofreading comments inline as part of this review
> 
>     * Let's have 1 more member of the team review this - @ilanolkies or @miltonwg can you add a PR review?

I'll review it!

Reviewer 3: I can add my review too, after this corrections

Author: I have just applied the @bguiz suggestions, thanks!

### Initialization
Initialize the `ERC20TransactionBuilder`.

```
const erc20TransactionBuilder = new ERC20TransactionBuilder(safe, ERC20Token)
```
Parameters:
- `safe: Safe` - a [Safe instance](https://github.com/gnosis/safe-core-sdk/blob/main/packages/safe-core-sdk/src/Safe.ts)
- `ERC20Token: Contract` - an [ethers.js Contract](https://docs.ethers.io/v5/api/contract/contract/) representing the ERC20 token

### Transfer transaction

```
const safeTransaction = await erc20TransactionBuilder.transfer(
  to,
  transfer
)
```
Parameters:
- `to: string` - the address will receive the amount of token specified with `transfer`
- `transfer: BigNumber` - the amount of token will be transferred to address `to`

### TransferFrom transaction

```
const safeTransaction = await erc20TransactionBuilder.transferFrom(
  from,
  to,
  value
)
```
Parameters:
- `from: string` - the address from which transfer the amount of token specified with `value`
- `to: string` - the address will receive the amount of token specified with `transfer`
- `value: BigNumber` - the amount of token will be transfered to address `to`

> >To execute the `transferFrom` successfully, the `from` address should explicitly authorize the safe account to spend an amount equal to or greater than `value`. Such operation can be authorized through the `approve` method of the `ERC20Token` Contract.

Reviewer: ```suggestion
> To execute the `transferFrom` successfully, the `from` address should explicitly authorize the safe account to spend an amount equal to or greater than `value`. Such operation can be authorized through the `approve` method of the `ERC20Token` Contract.

```

```
const fromToken = await MockERC20Token.connect(userFrom)
await fromToken.approve(safe.getAddress(), value)
```

### Approve transaction

```
const safeTransaction = await erc20TransactionBuilder.approve(
  spender,
  amount,
)
```
Parameters:
- `spender: string` - the address allowed to withdraw up to the amount of token specified with `amount`
- `amount: BigNumber` - the maximum amount allowed to be withdrawn

As stated in the [EIP20](https://eips.ethereum.org/EIPS/eip-20#methods) approve description:

> >"clients SHOULD make sure to create user interfaces in such a way that they set the allowance first to 0 before setting it to another value for the same spender."

Reviewer: ```suggestion

> clients SHOULD make sure to create user interfaces in such a way that they first set the allowance to 0, before setting it to another value for the same spender.
```

> For further details, please have a look at [the related document](https://docs.google.com/document/d/1YLPtQxZu1UAvO9cZ1O2RPXBbT0mooh4DYKjA_jp-RLM/)

Reviewer: ```suggestion
For further details, please read [*ERC20 API: An Attack Vector on Approve/TransferFrom Methods* by Mikhail Vladimirov and Dmitry Khovratovich](https://docs.google.com/document/d/1YLPtQxZu1UAvO9cZ1O2RPXBbT0mooh4DYKjA_jp-RLM/).

```

(let's make it clear that this is an external article)

layout: rsk
title: RIF Multisig SDK - transactions
---

**Required packages**:
- [@rsksmart/safe-transactions-sdk](https://github.com/rsksmart/safe-transactions-sdk)
- [@gnosis.pm/safe-core-sdk](https://github.com/gnosis/safe-core-sdk)

## Raw Transactions

> Raw transactions can be created by using directly the [safe-core-sdk](https://github.com/gnosis/safe-core-sdk) package or the [safe-transactions-sdk](https://github.com/rsksmart/).

Reviewer 2: >  [safe-transactions-sdk](https://github.com/rsksmart/).

we still don't have this repo published?

Author: We didn't when I started writing, but now I can add the link, thanks

```ts
const tx = await safe.createTransaction({
      to,
      data,
      value
    })
```

### safe-transactions-sdk

The `RawTransactionBuilder` provides the possibility to create raw transactions.

Initialize the `RawTransactionBuilder`.

```ts
const rawTransactionBuilder = new RawTransactionBuilder(safe)
```

Create the transaction.

```ts
const tx = await rawTransactionBuilder.rawTransaction(to, value, data)
```

## ERC20 Transactions

> The `ERC20TransactionBuilder` provides a set of methods related to [ERC20](https://eips.ethereum.org/EIPS/eip-20) transactions, including RIF token.

Reviewer 2: As we mention the RIF token, I'd add this link to it:
https://developers.rsk.co/rif/token/

### Initialization

> Initialize the `ERC20TransactionBuilder`.

Reviewer 2: shouldn't we mention that this is only possible by using the **safe-transactions-sdk**?

Author: I added an `import` statement to clarify the class is provided by the `safe-transaction-sdk` package.
